java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java

```diff
@@ -153,22 +166,29 @@ public void computedFrom(List<SymbolicValue> symbolicValues) {
     }
 
     @Override
-    public ProgramState setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
-      ProgramState result = programState;
+    public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
       if (leftOp.equals(rightOp)) {
-        return shouldNotInverse().equals(booleanConstraint) ? programState : null;
+        if (shouldNotInverse().equals(booleanConstraint)) {
+          return ImmutableList.of(programState);
+        }
+        return ImmutableList.of();
       }
-      result = copyConstraint(leftOp, rightOp, result, booleanConstraint);
-      if (result == null) {
-        return null;
+      List<ProgramState> result = copyConstraint(leftOp, rightOp, programState, booleanConstraint);
+      if (result.isEmpty()) {
+        return ImmutableList.of();
       }
-      result = copyConstraint(rightOp, leftOp, result, booleanConstraint);
-      if(result == programState) {
-        Map<SymbolicValue, Object> temp = Maps.newHashMap(programState.constraints);
-        temp.put(this, booleanConstraint);
-        return new ProgramState(programState.values, temp, programState.visitedPoints, programState.stack);
+      List<ProgramState> reversedResult = new ArrayList<>();
+      for (ProgramState ps : result) {
+        List<ProgramState> reversedStates = copyConstraint(rightOp, leftOp, ps, booleanConstraint);
+        if (reversedStates.size() == 1 && reversedStates.get(0) == programState) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)